### PR TITLE
DashboardGrid: Add support to filter panels using variable

### LINKED ISSUE #74891: Panel Filtering in Dashboard
Hi there. We moving over dashboards from an in-house legacy system to our Grafana instance. Something that's a bit unique about our current system is that users have created very large dashboards, with sometimes thousands of panels. This is possible to navigate because our in-house system provides a panel filter where a user can specify filtering regex and only display only those panel's whose title matches said regex. Finally, the regex is included in the URL so that a filtered view can be shared.

We realize this is an impractical paradigm and the obvious solution would be to force users to sub-divide or reorganize dashboards prior to moving to Grafana. However, due to the constraints of our migration to Grafana the only option is to do an initial copy-paste, followed by a dashboard refactor at our user's leisure.

I intend to submit a PR to add this panel regex filter feature and place it under a feature flag. However, I'd still like to ask for feedback to ensure the implementation is elegant and architecturally sound. With that I'd like to ask:

* Would it be more strategic to place this functionality into a new variable type? This way the [URL support](https://grafana.com/docs/grafana/latest/dashboards/build-dashboards/create-dashboard-url-variables/) would come for free
* If the variable route is not recommended, can you recommend a place to start? (e.g. New component, wiring to URL component, etc). I want to make sure I structure this neatly and not just duct-tape the feature in
* Could this feature be made more generic so that others can benefit? Our use case requires regex filtering but maybe we can make different types available (e.g. prefix filtering)

Thanks again and please forgive my shallow understanding of Grafana's internals.

**Why is this needed**:
- We require legacy dashboards to be moved as is
- These dashboards sometimes have thousands of panels
- We need to provide our users with a 1:1 panel filtering experience

**What would you like to be added**:
- A way to filter panels in a dashboard based on user inputed regex
- The regex should be encoded in the URL so that the state of the dashboard can be shared

**Who is this feature for?**
- For a large migration away from a legacy in-house solution to Grafana

### PR DESCRIPTION
**Pull request by jacobzelek**
**What is this feature?**

Users can add a special variable to their dashboards that will allow them to filter which panels are being displayed.

**Why do we need this feature?**

Only a small subset of users may need this feature. It's particular focus is for migrations that involve dashboards with hundreds, if not thousands of dashboards that require a fast way of drilling down.

**Who is this feature for?**

The feature was originally proposed to assists users during a migration from a legacy metrics platform. The details are mentioned in #74891.

**Which issue(s) does this PR fix?**:

Fixes #74891 

**Special notes for your reviewer:**

Please check that:
- [X] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard/dashgrid/DashboardGrid.tsx [hunk 1]
```diff
@@ -285,7 +362,7 @@ interface GrafanaGridItemProps extends React.HTMLAttributes<HTMLDivElement> {
   isViewing: boolean;
   windowHeight: number;
   windowWidth: number;
-  children: any;
+  children: any; // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (dprokop) [inline public/app/features/dashboard/dashgrid/DashboardGrid.tsx:365]: Why is this change here?

**Review by dprokop — COMMENTED** [review]
@s0lesurviv0r This looks really good, tested the behavior, and I think it works. The missing bit is the tests for the filtering behavior that I think would be good to have here.

The next step is to hide this feature behind a feature toggle. You can take some inspiration on how to create a feature toggle in this PR: https://github.com/grafana/grafana/pull/75560

File: public/app/features/dashboard/dashgrid/DashboardGrid.tsx (hunk 1 above)
Comment (s0lesurviv0r) [reply public/app/features/dashboard/dashgrid/DashboardGrid.tsx:365]: This kept failing the pre-hook linting. As far as I understand this was committed before that lint check was activated or someone was able to bypass the linter when committing this line.

**Comment by s0lesurviv0r** [comment]
Thanks for the feedback @dprokop. I'll put this behind the feature flag and I'll add said test cases.

**Comment by s0lesurviv0r** [comment]
@dprokop I've been exploring the best way to provide tests for this feature.

My first thought was to extend the component test suite in `DashGrid.test.tsx`. Quickly I realized that rendering `<DashboardGrid {...props} />` as is done in the existing test doesn't yield the expected HTML DOM. I figured out that the `AutoSizer` inside `DashGrid` isn't being populated when created via the test. I figured this may be because the component is rendered by itself and not within some sort of container.

I saw that the test for `PublicDashboardPage` in `PublicDashboardPage.test.tsx`, which has `DashGrid` as a child component, does successfully render the expected HTML DOM. This led me to believe that I should wrap the `DashGrid` component in a similar fashion to the `PublicDashboardPage` test. I've tried many different combinations that look approximately like this:
```
<GrafanaContext.Provider value={context}>
      <Provider store={store}>
        <Page>
          <DashboardGrid {...props} />
        </Page>
      </Provider>
    </GrafanaContext.Provider>
```

Nothing I've tried has made the `AutoSizer` render.

I'm wondering if you might have any advice on getting this working. I'd really like to have the entire `DashGrid` component rendered and wired into an event bus so it can be tested thoroughly. However, if this can't be done within the separate `DashGrid` test, can test this feature via the `PublicDashboardPage` test?

**Comment by dprokop** [comment]
@s0lesurviv0r -  I think the problem is the AutoSizer not being able to return a container width/height correctly. I think what you can do is mock the AutoSizer in your test file, similar to how it's done in `PublicDashboardPage`:

https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/containers/PublicDashboardPage.test.tsx#L37


Let me know if this doesn't work and we'll figure an alternative approach.

**Comment by s0lesurviv0r** [comment]
Nice! I feel foolish for missing that. I'll give it a try. Thanks!

**Comment by dprokop** [comment]
[s0lesurviv0r](https://github.com/s0lesurviv0r) I've pushed tests to your branch as we talked this morning. Any chance you sync your fork with main?

**Comment by s0lesurviv0r** [comment]
Hey @dprokop sorry this branch has become a mess. I rebased it and it pilled a lot of commits in. Let me work to fix this. The test cases look good though. Thanks!

**Comment by s0lesurviv0r** [comment]
Closed this PR and replacing with https://github.com/grafana/grafana/pull/77112
